Tools/muscand.1
.TH "MUSCAND" "1" "Release 0.1.13" "daelstorm" "Museek Daemon Plus"
.SH "NAME"
.LP 
\fBMuscand\fR \- File scanner daemon for Museek
.SH "SYNOPSIS"
.B muscand
[\-c <\fIfilename\fP>] [\-\-config <\fIfilename\fP>]
[\-b] [\-\-buddy]
[\-l] [\-\-list]
[\-h] [\-\-help]
.SH "DESCRIPTION"
.LP 
Muscand is a FAM service that scans paths already configured by \fImuscan\fP(1) or \fImusetup\fP(1). If it detects changes to the paths, it will update the Shares Database it is watching after a few seconds. There are two available shares databases, Normal and Buddies\-Only. Buddies\-Only is an optional shares database that, if enabled, is only accessable by the users you've chosen as "Buddies".
.LP 
Before running muscand, you will need a configured museekd,  which can be done with \fImusetup\fP(1) or fI>musetup\-gtk\fP(1), and add some shared paths with either of the setup tools or with \fImuscan\fP(1).
.SH "OPTIONS"
.LP 
Muscand accepts the following options:
.TP 
.B \-c <\fIfilename\fP>, \-\-config <\fIfilename\fP>
Use a different config file.
.TP 
.B \-b, \-\-buddy
Select Buddy\-Only Shares.
.TP 
.B \-h, \-\-help
Display Help, version and exit.
.SH "EXAMPLE"
.LP 
To run this program the standard way type:
.LP 
muscand
.LP 
Alternativly you use a different config file with:
.LP 
muscand \-\-config config.xml
.LP 
Or, you can watch Buddy\-Only shares:
.LP 
muscand \-\-buddy
.LP 
.SH "FILES"
.TP 
 \fI~/.museekd/config.xml\fR
The default location for the \fBmuseekd\fP config file.
.TP 
 \fI~/.museekd/config.shares\fR
The default location for the Normal shares database.
.TP 
 \fI~/.museekd/config.shares.state\fR
The default location for the active Normal shares database.
.TP 
 \fI~/.museekd/config.buddyshares\fR
The default location for the Buddy shares database.
.TP 
 \fI~/.museekd/config.buddyshares.state\fR
The default location for the active Buddy shares database's.
.SH "AUTHORS"
.LP 
Hyriand <hyriand@thegraveyard.org>
.LP 
daelstorm <daelstorm@gmail.com>
.SH "SEE ALSO"
.LP 
\fImucous\fP(1) \fImulog\fP(1) \fImurmur\fP(1) \fImuscan\fP(1) \fImuseekcontrol\fP(1) \fImuseekd\fP(1) \fImuseeq\fP(1) \fImusetup\fP(1) \fImusetup\-gtk\fP(1)


CHANGELOG
Museek Plus Changelog

0.1.13:  August 25th 2007

BUILD SYSTEMS
CMake, qmake, SCons and Python's distutils
 * RECOMMENDED: cmake
 * CMake compiles and installs everything. Read the INSTALL files for details.
 * qmake compiles just museeq. Read museeq/INSTALL
 * You can install the python bindings, apps, and clients with distutils.
   Run "python setup.py install" in each python sub-directory
 * SCons is still available, but not recommended. Read INSTALL.scons

MUSEEK DAEMON
 * Ignore SIGPIPE (to prevent crashes on BSDs)
 * Catch std::bad_alloc when unpacking a network message.
 * Always get IP from server, when requested
 * Autoretry downloads
 * Syslog support
 * Watch trusted users' status

MUSEEQ
 * Settings revamped
 * Shares configuration added to Settings
 * Added Audacioous to NowPlaying QSA script
 * Added colors to transfers' status column

MUCOUS
 * Setup revamped
 * Many, many bugfixes

MUSEEKCONTROL
 * Implemented a few more unfinished options

MURMUR
 * The basic features are here, enough for inclusion in this release.

0.1.12:  October 21th 2006

MUSEEK DAEMON
 * Fixed museekd not reacting on connect mode changes during run-time (hyriand)
 * Fixed the clear-initiating-transfer that causes a delayed crash (hyriand)
 * Character unpacking checks (fixes corrupted message crashes) (hyriand)
 * Trusted Users' status is now watched
 * IMPORTANT NOTE: the CFLAGS variable renamed to FLAGS and is now a String 
   instead of List, your mulocal.py will need to be modified or erased to allow
   SCons to run correctly. Not doing so will result in SCons failing at the 
   first compile-test.
 * Change the way the IP Address string is created to fix endian bug (hyriand) 
 * New SCons variable LOCKFLAGS disables SCons' automatic flag adding
 * Replaced depreciated call to Send Speed (34) on download completion
   with Send Upload Speed (121) on upload completion. (daelstorm)
 * Check for read/write permissions to config file before loading (daelstorm)
 * Added Global message (66) handler, and pass it to the clients (daelstorm)
 * Removed disconnect-on-hangup-signal (caused a loop)
 * Client message 'Download file to a directory' added (daelstorm)
 * Sends a Museek Daemon version message to clients on connection (daelstorm)
 * Sends the number of privileged users to clients on Server Login (daelstorm)

MUSCAN
 * Ignore .dotfiles and .dotdirs/
 * Removing trailing /s for shared directories
 
MUSEEQ
 * Fixed crash when disconnecting while still connecting to museekd by using
   "mSocket->deleteLater()" instead of "delete mSocket" (hyriand)
 * Fixed method of creating UNIX socket (hyriand)
 * Check option in settings to toggle display of exit dialog
 * Added nick's transfer list color patch transfers are now different colors,
   depending upon their state.
 * Added MUSEEQTRAYICON scons option to enable/disable trayicon at buildtime.
 * Moved Trayicon into Museeq class, made it change with away status and added
   menu option to toggle it.
 * Added /commands to private chat
 * Autoconnect to daemon option added
 * IP addresses can now be viewed in the log window
 * Updated Italian translation (requiem)
 * Updated Polish translation (belegdol)
 
MUSETUP-GTK
 * Scons tries more than one method to detect PyGTK
 
MUCOUS
 * The Setup Interface is now completely controllable from the keyboard
 * Browse Shares now has expandable/collapsable directories and resizeable
   windows. There are some other UI improvements, as well.
 * Added /me and status colors to Private chat 
 * Rewrote tab completion
 * Case-insensitive sorting
 * Refactored / Renamed lots of functions
 * Killed dozens of memory leaks
 * Removed vertical borders from private chat and rooms
 * Added scrolling ticker to chat rooms
 * Improved the Transfers colors and use vertical lines instead of pipes
 * Fixed some broken config options
 * Simplified creation of setup widgets
 * Setup's control over muscan is slightly better, attempts to detect museekd
   config, to better scan correct shares databases.
 * Hundreds of minor changes
 
0.1.11:  July 7th 2006

MUSEEK DAEMON
* ReAdd the server ping, but change the delay between them from 15 to 45 seconds
  to reduce load on the server and to not get banned :).
* Try to logout when shutting down
* Fixed the minPort == maxPort bug and maxPort > minPort bug, for peer
  connections. Previously, it'd skip the port listen attempt entirely. Now it
  tries the listen on minPort (if they equal) or on maxPort (if it's smaller 
  than minport) and  then gives up if it fails. 
* SCons can now set CCFLAGS with the CFLAGS list option
* Enable or disable installation of musetup-gtk with Scons bool option:
  MUSETUPGTK
* Choose which translations to install with the MUSEEQTRANSLATIONS Scons list
  option. (New translations may require modification to the SConstruct file)

MUSEEQ
* Added Timestamp toggle to the Settings menu
* Made user status appear in Private Chat (* user is away || * user is offline )
* Ported all .ui dialogs to cpp
* Added encodings comboboxes to settings dialog, as well as making the password
  dialog protected with ****s
* Added Translation *Support*
* Le_Vert translated museeq into French
* Requiem translated museeq into Italian
* Belegdol translated museeq into Polish
* SyphiliticVagina translated museeq into Spanish
* () translated museeq into German
* Senhor_das_Moscas translated museeq into Brazilian Portuguese
* Added more icons (ip, comments, alert, and small-icons for:  browse, userinfo,
  search, chatrooms, privatechat )
* Add two of mike-labo's themes from museeq. Read the README files for credits.

MUCOUS
* Version to 0.8.2
* New commands: /autoclear, /autoretry, /beep, /trustlist
* Add trusted users list to Users mode
* Make mucous behave more nicely when not connected to a museekd.
* Reorganize parts of setup mode, add new options for buddy-only shares, etc.
* Add timestamps to Debug log, bugfixes for browse mode, topbar variables
  merged into the new storage method


Tools/muscand.cc
#include <system.h>
#include <fam.h>

#include <Tools/scanner.hh>
#include <Muhelp/Muconf.hh>

#define MULOG_DOMAIN "Muscan.SC"
#include <Muhelp/Mulog.hh>

#include <iostream>

using std::string;
using std::map;
using std::cout;
using std::endl;
using std::cerr;
using std::vector;

class FAMDirScanner;

class FAMHandler
{
public:
	FAMHandler(const string& config_file, bool doBuddy);
	~FAMHandler();
	
	int load();
	int run();
	void save();
	
	void remove(FAMDirScanner *);
	
private:
	string shares, state;
	FAMConnection fc;
	vector<FAMDirScanner *> nodes, pending;
	time_t save_at;
	FAMDirScanner *root;
	
	void add(FAMDirScanner *);
	void exists(FAMDirScanner *ds, const char *filename, bool force_file);
	void deleted(FAMDirScanner *ds, const char *filename);
};

class FAMDirScanner : public DirScanner {
public:
	FAMDirScanner(FAMHandler *_fh, bool _f = true) : DirScanner(_f), fh(_fh) { };
	FAMDirScanner(FAMHandler *_fh, const string& _p) : DirScanner(_p), fh(_fh) { };
	~FAMDirScanner() { if(! fake) fh->remove(this); }
	
	DirEntry* new_folder(bool fake) { return new FAMDirScanner(fh, fake); }
	DirEntry* new_folder(const string& path) { return new FAMDirScanner(fh, path); }
	
	FAMHandler *fh;
	FAMRequest famRequest;
	
	bool isFake() { return fake; }
};

FAMHandler::FAMHandler(const string& config_file, bool doBuddy)
           : save_at(0), root(0)
{
	FAMCONNECTION_GETFD(&fc) = -1;

	Muconf config(config_file);
	if(! config.hasDomain("shares") || ! config["shares"].hasKey("database")) {
		cerr << "config file '" << config_file << "' incomplete or corrupt shares" << endl;
		exit(-1);
	}
	if(! config.hasDomain("buddy.shares") || ! config["buddy.shares"].hasKey("database")) {
		cerr << "config file '" << config_file << "' incomplete or corrupt buddy shares" << endl;
		exit(-1);
	}
	if (doBuddy)
		{
		string tmp = config["buddy.shares"]["database"];
		shares = tmp;
		state = shares + ".state";
		}
	else {
		string tmp = config["shares"]["database"];
		shares = tmp;
		state = shares + ".state";
		}

}

FAMHandler::~FAMHandler()
{
	if(root)
		delete root;

	if(FAMCONNECTION_GETFD(&fc) != -1)
		FAMClose(&fc);
}

int
FAMHandler::load()
{
	if(root)
		delete root;

	if(FAMCONNECTION_GETFD(&fc) != -1)
		FAMClose(&fc);
	
	if(FAMOpen2(&fc, "muscand") < 0)
	{
		cerr << "couldn't connect to FAM" << endl;
		return -1;
	}
	
	root = new FAMDirScanner(this);
	root->load(state);
	add(root);

	return 0;
}

void
FAMHandler::add(FAMDirScanner *h)
{
	if(find(nodes.begin(), nodes.end(), h) != nodes.end())
		cerr << "double insertion" << endl;
	else if(h->isFake())
		cerr << "it's fake.." << endl;
	else
	{
		if(! h->path.empty())
		{
			cerr << "Registering " << h->path << endl;
			pending.push_back(h);
		}
	}
	
	map<string,DirEntry*>::iterator it = h->folders.begin(), end = h->folders.end();
	for(; it != end; ++it)
		add((FAMDirScanner*)(*it).second);
}

void
FAMHandler::remove(FAMDirScanner *h)
{
	if(h->isFake())
	{
		cerr << "trying to remove a fake entry" << endl;
		return;
	}
	
	vector<FAMDirScanner*>::iterator it = find(nodes.begin(), nodes.end(), h);
	if(it == nodes.end())
		cerr << "Possible corruption (removal of non-existing entry)" << endl;
	else
	{
		FAMCancelMonitor(&fc, &h->famRequest);
		nodes.erase(it);
	}
	
	vector<FAMDirScanner *>::iterator it2 = find(pending.begin(), pending.end(), h);
	if(it2 != pending.end())
	{
		cerr << "Removing pending registration" << endl;
		pending.erase(it2);
	}
}

int
FAMHandler::run()
{
	int fam_fd = FAMCONNECTION_GETFD(&fc);
	
	while(1)
	{
		if(save_at && time(NULL) >= save_at)
		{
			cerr << "saving updated shares database" << endl;
			save();
			save_at = 0;
		}
		
		fd_set rfds, wfds;
		struct timeval tv;
		
		FD_ZERO(&rfds);
		FD_ZERO(&wfds);
		FD_SET(fam_fd, &rfds);
		if(! pending.empty())
			FD_SET(fam_fd, &wfds);
		
		tv.tv_sec = 5;
		tv.tv_usec = 0;
		
		int retval = select(fam_fd+1, &rfds, &wfds, 0, &tv);
		if(retval == -1)
		{
			FAMClose(&fc);
			cerr << "Select error on FAM socket" << endl;
			return -1;
		}
		if(retval == 0)
			continue;
		
		if(FD_ISSET(fam_fd, &wfds) && pending.size())
		{
			vector<FAMDirScanner *>::iterator it = pending.begin();
			FAMMonitorDirectory(&fc, (*it)->path.c_str(), &(*it)->famRequest, *it);
			nodes.push_back(*it);
			pending.erase(it);
		}
		
		if(FD_ISSET(fam_fd, &rfds))
		{
			while(FAMPending(&fc))
			{
				FAMEvent fe;
				if(FAMNextEvent(&fc, &fe) < 0)
				{
					FAMClose(&fc);
					cerr << "Read error on FAM socket" << endl;
					return -1;
				}
				
				if(fe.code == FAMExists || fe.code == FAMCreated || fe.code == FAMChanged) {
					exists((FAMDirScanner *)fe.userdata, fe.filename, fe.code == FAMChanged);
				} else if(fe.code == FAMDeleted) {
					deleted(static_cast<FAMDirScanner*>(fe.userdata), fe.filename);
				} else
					cout << "ping " << fe.code << fe.filename << endl;
			}
		}
	}
}

void
FAMHandler::save()
{
	root->save(state);

	DirEntry folded;

	root->fold(&folded);
	folded.save(shares);

}

void
FAMHandler::exists(FAMDirScanner *ds, const char *filename, bool force_file)
{
	std::string path = ds->path + "/" + filename;
	
	if(ds->path == filename || find(nodes.begin(), nodes.end(), ds) == nodes.end())
		return;
	
	struct stat s;
	if(stat(path.c_str(), &s) == -1)
		return;
	
	if(S_ISREG(s.st_mode))
	{
		if(force_file || ds->files.find(filename) == ds->files.end())
		{
			cout << "new file: " << path << endl;
			FileEntry fe = ds->scan_file(path);
			fe.size = s.st_size;
			ds->files[filename] = fe;
			save_at = time(0) + 15;
		}
	}
	else if(S_ISDIR(s.st_mode))
	{
		if(ds->folders.find(path) == ds->folders.end())
		{
			cout << "scanning: " << path << endl;
			ds->add(path);
			FAMDirScanner *ns = static_cast<FAMDirScanner *>(ds->folders[path]);
			ns->scan();
			add(ns);
			save_at = time(NULL) + 15;
		}
	}
}

void
FAMHandler::deleted(FAMDirScanner *ds, const char *filename)
{
	if(ds->path == filename)
	{
		cout << "hard deleting " << ds->path << " -- " << filename << endl;
		int ix = ds->path.rfind('/');
		string parent = ds->path.substr(0, ix), child = ds->path.substr(ix + 1);
		vector<FAMDirScanner *>::iterator it = nodes.begin(), end = nodes.end();
		for(; it != end; ++it)
		{
			if((*it)->path == parent)
			{
				cout << "found it" << endl;
				(*it)->folders.erase(filename);
				delete ds;
				save_at = time(0) + 15;
				return;
			}
		}
		cout << parent << " -- " << child << endl;
		return;
	}
	else
		ds->files.erase(filename);
}

void help() {
	cout << "muscand [-c --config PATH] [-b --buddy] [-h --help]" << endl;
	cout << "Version 0.1.13" << endl;
	exit(-1);
}

int main(int argc, char **argv)
{
#ifdef RELAYED_LIBFAM
	extern int libfam_is_present;
	if(! libfam_is_present)
	{
		cerr << "libfam not found, aborting" << endl;
		return -1;
	}
#endif
	string config_file = string(getenv("HOME")) + "/.museekd/config.xml";
	bool doBuddy = false;
	for(int i = 1; i < argc; i++) {
		string arg = argv[i];
		if(arg == "-c" || arg == "--config") {
			++i;
			config_file = string(argv[i]);
		} 
		else if(arg == "-b" || arg == "--buddy") {
			++i;
			doBuddy = true;
		} 
		else if(arg == "-h" || arg == "--help") {
			++i;
			help();
			
		}
	}
	
	FAMHandler fh(config_file, doBuddy);
	if(fh.load())
		return -1;
	fh.run();
	
	return 0;
}
